DisguiseUnityRenderStream/Runtime/DisguiseRenderStreamWithClustercs.cs

```diff
@@ -150,81 +156,180 @@ public static ClusterParams ProcessClusterParams(ClusterParams clusterParams)
-                clusterParams.MulticastAddress = (string)clusterKey.GetValue("MulticastAddress");
-                clusterParams.Port = (int)clusterKey.GetValue("MulticastPort");
-                clusterParams.AdapterName = (string)clusterKey.GetValue("AdapterName");
+                clusterParams.NodeID = NegotiateNodeID(clusterParams);
+                ClusterDebug.Log($"Auto-assigning node ID {clusterParams.NodeID} (repeaters: {clusterParams.RepeaterCount})");
+
+                clusterParams.Fence = FrameSyncFence.External;
+                clusterParams.ClusterLogicSpecified = true;
+
+                // Arbitrarily assign Node 0 as the emitter
+                clusterParams.EmitterSpecified = clusterParams.NodeID == 0;
             }
-            else if (maybeRepeaterCount.HasValue)
+            catch (Exception e)
             {
-#endif
-                clusterParams.RepeaterCount = maybeRepeaterCount.Value;
+                Debug.LogException(e);
+                clusterParams.ClusterLogicSpecified = false;
+            }
+
+            return clusterParams;
+        }
 
-                // If we're running several nodes on the same machine, use a single-access file to keep track
-                // of node ids in use.
-                var retries = 0;
-                var nodeIdFilePath = Path.Combine(Path.GetTempPath(), Application.productName + ".node");
-                while (!detectedNodeId && retries < 10)
+        /// <summary>
+        /// Negotiate a distinct ID amongst other running render nodes. Uses the same multi-cast end point
+        /// as the core Cluster Display logic.
+        /// </summary>
+        /// <param name="clusterParams">Parameters containing the networking information.</param>
+        /// <returns></returns>
+        /// <exception cref="TimeoutException">The operation timed out.</exception>
+        /// <exception cref="OperationCanceledException">Multiple errors were encountered. See the log for details.</exception>
+        static byte NegotiateNodeID(ClusterParams clusterParams)
+        {
+            var address = IPAddress.Parse(clusterParams.MulticastAddress);
+            var sendEndPoint = new IPEndPoint(address, clusterParams.Port);
+            var timeoutMilliseconds = (int)clusterParams.HandshakeTimeout.TotalMilliseconds;
+            
+            using var udpClient = new UdpClient();
+            udpClient.EnableMulticast(address,
+                clusterParams.Port,
+                clusterParams.AdapterName,
+                timeoutMilliseconds);
+
+            int nodeIdResult;
+            var expectedNodeCount = clusterParams.RepeaterCount + 1;
+            
+            // Message that indicates that we're available to join a group.
+            // Each node must have a unique announcement message.
+            var announceMessage = $"announce:{Environment.MachineName}{Process.GetCurrentProcess().Id}";
```

Comment:
Brainstorm / Minor: How confident are we that the machine name is unique?  Although that wouldn't the best network setup, we could have two windows machine with the same name on the network and TCP/UDP would still be working just fine.

Do we want to use the IP address of the network adapter we are using to do the negotiation?  This IP has to be unique otherwise the network simply does not work :)

Reply:
I hadn't considered the possibility that machine names aren't unique. Good catch.